### DISCUSSION TITLE
How do I manage my `/etc` folder using chezmoi?

### DISCUSSION THREAD

**Discussion by ayushnix** [twpayne/chezmoi#1510] [Q&A] [answered]
I'm looking for recommendations on how to manage my `/etc` folder using chezmoi. I know there's [etckeeper](https://etckeeper.branchable.com/) but I'd rather use chezmoi for this since I'm already using it for my user dotfiles.

Right now, I'm using a simple [script](https://github.com/ayushnix/dotfiles/blob/master/private_dot_local/private_bin/executable_etcopy) to [periodically](https://github.com/ayushnix/dotfiles/blob/master/private_dot_config/systemd/user/etcopy.timer) copy some directories and files from `/etc` into my `XDG_CONFIG_HOME` folder and then manage them using chezmoi.

Is there a "better" way that I can manage them, preferably without executing chezmoi as the root user or changing permission of `/etc`?

**Comment by felipecrs** [comment, marked as answer]
I think that running as root would be a need, otherwise how could chezmoi write to /etc? In my dotfiles, I currently configure the files under `/etc` with `sudo` inside of my `run_` scripts. An example:

https://github.com/felipecrs/dotfiles/blob/master/home/run_once_before_1_setup_apt_packages.sh.tmpl#L68-L74

It may not be the cleanest approach, neither the most performant and bullet proof, but it works for me.

**Reply by ayushnix** [reply]
> otherwise how could chezmoi write to /etc

I guess I'd be okay if `chezmoi apply` doesn't work as intended (copying `/etc` dotfiles from chezmoi repo to my system `/etc`). I can then copy them manually using a script.

**Reply by ayushnix** [reply]
Using `run_once_` scripts sounds fine for now. I'll look at Ansible as well as suggested by twpayne. Thanks!

**Reply by Lockszmith-GH** [reply]
check out [comtrya](https://github.com/comtrya/comtrya#readme) - might be more suited to what you're asking than ansible.

**Reply by ronlut** [reply]
@felipecrs I see the file you referenced disappeared. Did you find a better way to do it? Care to share how you do it today? Thanks

**Reply by felipecrs** [reply]
That's a long gone approach. But here is the permalink to that file: https://github.com/felipecrs/dotfiles/blob/fb86ab241655d4991734fd7981780578b625f335/home/run_once_before_1_setup_apt_packages.sh.tmpl#L68-L74

I now use something that I call "rootmoi". I've got no time to explain it now (but I believe I did somewhere, although I can't find it).

Hopefully you will understand by checking these references:

https://github.com/felipecrs/dotfiles/blob/c3e0e400e544b3918a856cb10b55672c6ab4c827/home/dot_local/bin/executable_rootmoi.tmpl
https://github.com/felipecrs/dotfiles/blob/c3e0e400e544b3918a856cb10b55672c6ab4c827/home/.chezmoiscripts/run_after_20-run-rootmoi.sh.tmpl
https://github.com/felipecrs/dotfiles/tree/c3e0e400e544b3918a856cb10b55672c6ab4c827/root

I find it pretty neat. :)

**Reply by 00sapo** [reply]
> That's a long gone approach. But here is the permalink to that file: https://github.com/felipecrs/dotfiles/blob/fb86ab241655d4991734fd7981780578b625f335/home/run_once_before_1_setup_apt_packages.sh.tmpl#L68-L74
> 
> I now use something that I call "rootmoi". I've got no time to explain it now (but I believe I did somewhere, although I can't find it).
> 
> Hopefully you will understand by checking these references:
> 
> https://github.com/felipecrs/dotfiles/blob/c3e0e400e544b3918a856cb10b55672c6ab4c827/home/dot_local/bin/executable_rootmoi.tmpl https://github.com/felipecrs/dotfiles/blob/c3e0e400e544b3918a856cb10b55672c6ab4c827/home/.chezmoiscripts/run_after_20-run-rootmoi.sh.tmpl https://github.com/felipecrs/dotfiles/tree/c3e0e400e544b3918a856cb10b55672c6ab4c827/root
> 
> I find it pretty neat. :)

Hey, thanks! I used your setup and it works! I think we could make a chezmoi "extension" out of it...

**Reply by felipecrs** [reply]
I will explore this option.

**Reply by DrummyFloyd** [reply]
i maanged to make it  work on my own dotfiles, and it work like a charm thank you @felipecrs

**Comment by twpayne** [comment]
This is straying into the territory of [managing files outside your home directory](https://github.com/twpayne/chezmoi/blob/master/docs/FAQ.md#can-i-use-chezmoi-to-manage-files-outside-my-home-directory) - which is an explicit non-goal for chezmoi. You might be better off with Ansible/Chef/Puppet/Salt.

**Reply by ayushnix** [reply]
The suggestion about using `run_once_` scripts sounds ok for now. I'll create a `run_once` script that copies from chezmoi's git repo to `/etc`. I'll take a look at Ansible as well for this. Thanks!

**Reply by someonewithpc** [reply]
@twpayne That link is dead and seems to be referenced in multiple places, but I can't find where it's meant to point. This project looks great, but not being easily able to manage files outside $HOME seems like a major limitation. I assume that link has reasoning on why, so I can't say it's not reasonable, though

**Reply by bradenhilton** [reply]
@someonewithpc https://www.chezmoi.io/user-guide/frequently-asked-questions/design/#can-i-use-chezmoi-to-manage-files-outside-my-home-directory

**Comment by nandalopes** [comment]
@ayushnix You can take a look on my branch **friends**, where I manage install profiles and run_once scripts if a certain file changes:

- [install profiles](https://github.com/nandalopes/dotfiles/blob/friends/root/.chezmoi.yaml.tmpl#L79)
- [list of apt packages](https://github.com/nandalopes/dotfiles/blob/friends/root/private_dot_backup/install/apt/modify_install.tmpl)
- [script that adds ppa sources](https://github.com/nandalopes/dotfiles/blob/friends/root/.chezmoiscripts/10_scripts/run_onchange_after_apt_sources_ppa.sh.tmpl)
- [apt install script](https://github.com/nandalopes/dotfiles/blob/friends/root/.chezmoiscripts/30_scripts/run_onchange_after_apt_install.sh.tmpl)

**Comment by mwoz123** [comment]
```
 Answered by [felipecrs](https://github.com/felipecrs)
[on Oct 9, 2021](https://github.com/twpayne/chezmoi/discussions/1510#discussioncomment-1453446)
I think that running as root would be a need, otherwise how could chezmoi write to /etc? In my dotfiles, I currently configure the files under /etc with sudo inside of my run_ scripts. An example:

https://github.com/felipecrs/dotfiles/blob/master/home/run_once_before_1_setup_apt_packages.sh.tmpl#L68-L74
```
Marked as answer link doesn't exist anymore :(

Updated link from @felipecrs  dotfiles:  [run_once_before_1_setup_apt_packages.sh.tmpl#L68-L74](https://github.com/felipecrs/dotfiles/blob/fb86ab241655d4991734fd7981780578b625f335/home/run_once_before_1_setup_apt_packages.sh.tmpl#L68-L74)

**Comment by nandalopes** [comment]
There's a great tool to manage /etc, [etckeeper](https://etckeeper.branchable.com/)

**Comment by mwoz123** [comment]
I know **it's not recommended as it put security at risk,** but I managed to create such chezmoi config. 

If you don't care about security - my alias from `.bashrc` :

```bash
alias schezmoi="sudo chezmoi --destination /etc --source /etc/chezmoi/data --working-tree /etc/chezmoi/data --config /etc/chezmoi/chezmoi.toml"
```
then `schezmoi init ...` etc

but once again **be aware of security risks** e.g. here simple `cd` gives root shell:

```bash
marcin@pc:~$ schezmoi status
[sudo] password for marcin: 
 M root/.bashrc
marcin@pc:~$ schezmoi cd
root@pc:/etc/chezmoi/data# whoami
root
root@pc:/etc/chezmoi/data# 
```
ouch! 

you have been warned ;)

**Reply by felipecrs** [reply]
Out of curiosity, if `chezmoi cd` is the only flaw, it can be fixed by using a shim to filter out this option.

But, in my case, I always use `sudo` without passwords for my user so this isn't an additional exposure for me. ;)

**Comment by SilverRainZ** [comment]
Sorry for necro-bump :-X

Heavily inspired by this discussion, I built a simple chezmoi wrapper called chezetc for managing files under `/etc`.

Its usage is identical to chezmoi because it just forwards all flags to chezmoi. It is quite simple and complete to handle the permission issue. For example, `chezetc cd`, `chezetc edit`, and `chezetc apply` will not expose any root shell.  

Please checkout https://github.com/SilverRainZ/chezetc for more details. I hope it is helpful for some guys.

**Reply by twpayne** [reply]
Wow, very cool!

**Reply by DrummyFloyd** [reply]
cool, will may have look, even if the acutal implementation workaround `rootmoi` work as expected

maybe add some advise to use it as `chezmoi plugins` tho'

thank for your work

**Reply by 00sapo** [reply]
Same here. It's nice that chezetc is more structured project than rootmoi, but rootmoi doesn't require additional commands, from the user perspective, it's just an added ability of Chezmoi (except for re-add). What I mean is that the user can just run `chezmoi update` or `chezmoi apply` and the / files updated/applied as well. For `re-add`, I don't remember why but my config is not supporting it

**Reply by felipecrs** [reply]
You can probably achieve a similar thing by running chezetc as a chezmoi script though.
